doc/sfputil/read_write_eeprom_by_page.md

```diff
@@ -0,0 +1,253 @@
+# sfputil | add the ability to read/write any byte from EERPOM both by page and offset #
+
+## Table of Content
+
+### Revision
+
+### Scope
+
+This document describes SONiC feature to read/write cable EEPROM data via SONiC CLI.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+CLI sfputil shall be extended to support reading/writing cable EEPROM by page and offset. This implementation is based on existing platform API `sfp.read_eeprom` and `sfp.write_eeprom`.
+
+### Requirements
+
+- Support reading/writing cable EEPROM data by page, offset and size. For sff8472, wire address "a0h" or "a2h" must be provided by user.
+- Support basic validation for input parameter such as page, offset and size.
+- Support reading/writing cable EEPROM for all types of cables except RJ45.
+- Support non flat memory mode reading/writing. In this mode, user shall provide page and offset according to the standard. For example, CMIS page 1h starting offset is 128, offset less than 128 shall be treated as invalid.
+- Support flat memory mode reading/writing (Phase 2). In this mode, the EEPROM memory shall be treated as continues flat memory. This design document will not discuss this mode.
+- Vendor who does not support `sfp.read_eeprom` and `sfp.write_eeprom` is expected to raise `NotImplementedError`, this error shall be properly handled
+- Others error shall be treated as read/write failure
+- Vendor specific implementation is not required but still possible.
+
+
+### Architecture Design
+
+The current architecture is not changed.
+
+### High-Level Design
+
+Submodule sonic-platform-common and sonic-utilities shall be extended to support this feature. Vendor specific implementation is not required but still possible.
+
+#### sonic-platform-common
+
+The existing API `sfp.read_eeprom` and `sfp.write_eeprom` accept "overall offset" as parameter. They have no concept of page. If user wants to use it, user has to convert page and offset to "overall offset" manually. Different cable types use different covert method according to the standard. So, it is not user friendly to provide such API directly to user. Two new API shall be added to support this feature:
+
+```python
+
+def read_eeprom_by_page(self, page, offset, size, wire_addr=None, flat=False):
+    """
+    Read EEPROM by page
+
+    Args:
+        page: EEPROM page number. Raise ValueError for invalid page.
+        offset: EEPROM page offset. Raise ValueError for invalid offset.
+        size: Number of bytes to be read. Raise ValueError for invalid size.
+        wire_addr: Wire address. Only valid for sff8472. Raise ValueError for invalid wire address.
+        flat: Read mode.

```

Comment:
Why is this parameter germane?  Whether the module is flat (or not) should be determined by the advertised MemoryModel indicator at page0 byte2 bit7.  Why would it be desirable for the operator (user of sfputil) to override this indicator when reading/writing?   Same question for write_eeprom_by_page...

Reply:
Hi, the `flat` flag means that whether we should treat the EEPROM as a continues memory. It is not used currently. So, to avoid confusion, I will remove this flag since it is not used in this phase.